src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientFactory.java

```diff
@@ -43,7 +45,7 @@
  * instance.
  */
 @ThreadSafe
-public interface MongoClientFactory {
+public interface MongoClientFactory extends Serializable {
```

Comment:
I found the answer regarding `MongoConfig` being `Serializable`: 

1. `org.apache.spark.sql.connector.write.DataWriterFactory` is `Serializable`, so must be `MongoDataWriterFactory`.
2. `MongoDataWriterFactory` refers to `WriteConfig`.

`RowToBsonDocumentConverter` and `InternalRowToRowFunction` must be `Serializable` for the same reason.

@rozza This leaves me with a question about `MongoClientFactory`: isn't it better to make `AbstractMongoConfig.mongoClientFactory` `transient` than to make `MongoClientFactory` `Serializable`?

Reply:
The cache is not `Serializable`, each process creates and populates its cache from scratch. Thus, the fact that `MongoClientFactory` is a key in `MongoClientCache` does not impose the serializability requirement on the `MongoClientFactory` interface.

> it can be sent to each executor node and only requires a single instance

If `AbstractMongoConfig.mongoClientFactory` is made transient, each `AbstractMongoConfig` (whether deserialized or not) will still create not more than one `MongoClientFactory`. Having one `MongoClientFactory` per `AbstractMongoConfig` is achieved by the method `AbstractMongoConfig.getMongoClientFactory`, which is agnostic of the serializability of the `MongoClientFactory` objects.

> I don't see a benefit to make it transient.

The benefit is that `MongoClientFactory` will not be `Serializable`, which reduces the space of things one has to care about. And, importantly, there are no drawbacks. One could also say that this improves the performance of serializing/deserializing `AbstractMongoConfig` objects and reduces the size of its serialized form, but the improvement is so negligible and irrelevant, that such an argument is vanishingly weak, and I am pointing it out just for the sake of completeness.

Reply:
Done